### ISSUE TITLE
Semantic Highlighting

### ISSUE THREAD

**Issue by JeanMertz** [rust-analyzer/rust-analyzer#604]
rust-analyzer [supports highlighting](https://github.com/rust-analyzer/rust-analyzer/commit/9863b9161d8d702848516be70c5c8161b7f382e8) based on a static list of decoration objects _(limiting the flexibility of the current implementation)_, and mapping them to a [static set of color codes](https://github.com/rust-analyzer/rust-analyzer/blob/1d7baef0e96d6c1db2d049f3770588f07b8d739f/editors/code/src/highlighting.ts#L22-L32) _(limiting the usefulness for people with other color preferences)_.

There is a [proposal to add a semantic highlighting protocol extension](https://github.com/Microsoft/vscode-languageserver-node/pull/367) to LSP, which would make this feature much more powerful.

To get an idea of what's possible with this new extension, see [this comment](https://github.com/Microsoft/vscode-languageserver-node/pull/367#issuecomment-411043236) with some gifs, showing this in action.

I'm opening this issue to get some visibility on this, to start tracking the state of that extension, and perhaps get a discussion going on the best way forward.

I'm fairly new to this project, but I've recently switched from RLS to rust-analyzer for my Rust development needs, and am rooting for this project to bring a _much_ better IDE story to Rust in the near future. I also hope to be able to contribute back to the project in the near future.

**Comment by DJMcNab** [comment]
I think the plan has always been to use the semantic highlighting due to appear in the LSP. However, when we made our current extension, the correct apis were not available.

My understanding is that we want to wait until that extension is integrated into the protocol before we start depending on it, so that rust analyser can work on vscode stable.

If that extension is in a state where the Api is stable, I'm sure we won't reject an implementation using it, but it probably isn't worth switching at the moment.

**Comment by matklad** [comment]
The core problem here is that VS Code itself does not have an API to do custom syntax highlighting. The decorations we currently use are for another purpose.

**Comment by JeanMertz** [comment]
> The core problem here is that VS Code itself does not have an API to do custom syntax highlighting

I'm confused. If VS Code does not support custom syntax highlighting, then how can rust-analyzer support its own syntax highlighting in VS Code?

Also, I'm assuming that if the extension to the LSP adds support for this, VS Code will probably not be far behind in actually supporting use-cases like these.

I wasn't implying that we should implement this _now_, but I did want to get some eyes on this (I couldn't find any other discussion mentioning this LSP extension), to see what we could do in the future, when this extension stabilises.

**Comment by ice1000** [comment]
> I'm confused. If VS Code does not support custom syntax highlighting, then how can rust-analyzer support its own syntax highlighting in VS Code?

An issue for this: microsoft/vscode#585

**Comment by JeanMertz** [comment]
Thanks @ice1000, that issue also links to https://github.com/microsoft/vscode/issues/77133 with quite some recent action.

Seems like this is pretty close to being supported natively in VSCode.

**Comment by ice1000** [comment]
> Seems like this is pretty close to being supported natively in VSCode.

I hope so. Yes, I *hope* so.

**Comment by kjeremy** [comment]
#2655 provides the LSP support for this.

**Comment by matklad** [comment]
@kjeremy wow, did they already added this as a proposed API? We should try it ought! I though that they are only planing to add this to VS Code...

**Comment by lnicola** [comment]
See the proposal [here](https://github.com/microsoft/vscode-languageserver-node/blob/5a9b33c23de84c3341e011e79221795a8059375b/protocol/src/protocol.semanticHighlighting.proposed.md). It's still using TM scopes, although I don't quite understand how the scopes are represented.

CC @seivan

**Comment by matklad** [comment]
Ah, so that proposal is not yet merged into mainline spec (even as a proposed addition), and so is not supported by VS Code (but is supported by kak and eclipse theia). We might want to support this, but I expect that the eventual proposal will be modified heavily (but I might be wrong!).

**Comment by kjeremy** [comment]
Ah that's a disappointment.

**Comment by seivan** [comment]
@lnicola Thanks for the ping, now correct me if I am wrong, but would this let us map TM scopes to parts of the editor? 
E.g we could map existing tokens to the range that maps `fn func(arg: &str)`?
That means I could get rid of `scope.ts` but still need `scope_mapper.ts` that would map the rust scopes to the common TM scopes?

**Comment by kjeremy** [comment]
This is in the plan vscode's [January 2020 iteration](https://github.com/microsoft/vscode/issues/87479)

**Comment by kjeremy** [comment]
The semantic token api is in preview here: https://code.visualstudio.com/updates/v1_41#_semantic-tokens-provider

An example of this API integrated with a java LSP implementation: https://github.com/microsoft/vscode/issues/86415

**Comment by benmp** [comment]
#3159 
#3321 
#3322

**Comment by matklad** [comment]
Indeed, thanks @benmp !
